axon/threads_test.go

```diff
@@ -0,0 +1,353 @@
+	netA, netB := buildIdenticalNetworks(t, pats, 1, 1, 1)
+
+	fun := func(net *Network, ltime *Time) {
+		net.Cycle(ltime)
+	}
+
+	// by splitting the epochs into three parts for netB, we make sure that the
+	// training is fully deterministic and not dependent on the `rand` package,
+	// as we re-set the seed at the beginning of runFunEpochs.
+	runFunEpochs(pats, netB, fun, 1)
+	runFunEpochs(pats, netB, fun, 2)
+	runFunEpochs(pats, netA, fun, 10)
+	runFunEpochs(pats, netB, fun, 7)
+
+	// compare the resulting networks
+	assertNeuronsEqual(t, netA, netB)
+
+	// sanity check, to make sure we're not accidentally sharing pointers etc.
+	assert.True(t, neuronsAreEqual(netA, netB))
+	runFunEpochs(pats, netA, fun, 1)
+	assert.False(t, neuronsAreEqual(netA, netB))
+}
+
+func TestMultithreadedSendSpike(t *testing.T) {
+	pats := generateRandomPatterns(10)
+	// run an absurd amount of threads, to make sure we encounter race conditions
+	// goroutines are cheap, they barely cost any memory
+	netS, netM := buildIdenticalNetworks(t, pats, 1, 16, 1)
+
+	fun := func(net *Network, ltime *Time) {
+		net.Cycle(ltime)
+	}
+
+	runFunEpochs(pats, netM, fun, 2)
+	runFunEpochs(pats, netS, fun, 2)
+
+	// compare the resulting networks
+	assertNeuronsEqual(t, netS, netM)
+
+	// sanity check, to make sure we're not accidentally sharing pointers etc.
+	assert.True(t, neuronsAreEqual(netS, netM))
+	runFunEpochs(pats, netS, fun, 1)
+	assert.False(t, neuronsAreEqual(netS, netM))
+}
+
+func TestMultithreadedNeuronFun(t *testing.T) {
+	pats := generateRandomPatterns(10)
+	netS, netM := buildIdenticalNetworks(t, pats, 16, 1, 1)
+
+	fun := func(net *Network, ctime *Time) {
+		net.Cycle(ctime)
+	}
+
+	runFunEpochs(pats, netM, fun, 3)
+	runFunEpochs(pats, netS, fun, 3)
+
+	// compare the resulting networks
+	assertNeuronsEqual(t, netS, netM)
+
+	// sanity check, to make sure we're not accidentally sharing pointers etc.
+	assert.True(t, neuronsAreEqual(netS, netM))
+	runFunEpochs(pats, netS, fun, 1)
+	assert.False(t, neuronsAreEqual(netS, netM))
+}
+
+func TestMultithreadedSynCa(t *testing.T) {
+	pats := generateRandomPatterns(10)
+	netS, netM := buildIdenticalNetworks(t, pats, 16, 1, 16)
+
+	fun := func(net *Network, ltime *Time) {
+		net.Cycle(ltime)
+	}
+
+	runFunEpochs(pats, netM, fun, 3)
+	runFunEpochs(pats, netS, fun, 3)
+
+	// compare the resulting networks
+	assertNeuronsEqual(t, netS, netM)
+
+	// sanity check, to make sure we're not accidentally sharing pointers etc.
+	assert.True(t, neuronsAreEqual(netS, netM))
+	runFunEpochs(pats, netS, fun, 1)
+	assert.False(t, neuronsAreEqual(netS, netM))
+}
+
+// assert that neuron fields are equal. We could also check for Synapse equality,
+// but over time different Synapse weights will lead to different Neuron fields.
+func assertNeuronsEqual(t *testing.T, netS *Network, netM *Network) {
+	maxDiff := struct {
+		diff      float64
+		field     string
+		neuronIdx int
+		valShould float64
+		valIs     float64
```

Comment:
I'd name these two fields to correspond to the two Network arguments so it's clearer how they relate.
E.g., `valS` and `valM`